optuna/distributions.py

```diff
@@ -301,6 +312,26 @@ def _contains(self, param_value_in_internal_repr):
         index = int(param_value_in_internal_repr)
         return 0 <= index and index < len(self.choices)
 
+    @staticmethod
+    def _to_valid_choice(choice):
+        # type: (Any) -> Union[None, bool, float, str]
+
+        if choice is None or isinstance(choice, (bool, float, str)):
+            return choice
+        try:
+            choice = float(choice)
+        except TypeError:
+            message = (
+                "Choices for a categorical distribution should be a tuple of None, bool, float, "
+                "str and castable to float for persistent storage but contains {} which is of "
+                "type {}.".format(choice, type(choice).__name__))
+            warnings.warn(message)
+
+            logger = logging._get_library_root_logger()
+            logger.warning(message)
```

Comment:
It might be the case that we only want warn once but it's not obvious how. For instance, it could be once during the process lifetime or once per choice configuration. Where that logic is to be implemented is I guess also open for discussion (e.g. in the logging utilities using [logging.captureWarnings](https://docs.python.org/3.5/library/logging.html#integration-with-the-warnings-module)). I left it out but am very open for ideas.

Reply:
I agree with you. Since it is noisy for users to emit the same warning repeatedly, I want a function like `logging.warning_once(...)`. However, I think that this topic should be discussed in another issue to keep this PR simple.